### PR TITLE
Add long-running endowment permission

### LINKED ISSUE #430: Add long-running snap permission
A proposed solution to the timeout problem #357.

To solve the problem, we will add a specific permission that snaps can request to disable timeouts for their execution.

### PR DESCRIPTION
**Pull request by rekmarks**
Adds a permission for endowing snaps with infinite request processing time.

Fixes #430

### CODE REVIEW

**Review by ritave — DISMISSED** [review]
Lacks implementation

**Comment by FrederikBolding** [comment]
> Lacks implementation

Added 😉

File: packages/controllers/src/snaps/SnapController.ts [hunk 1]
```diff
@@ -1762,7 +1765,18 @@ export class SnapController extends BaseController<
...
+    const isLongRunning = this.messagingSystem.call(
+      'PermissionController:hasPermission',
+      snapId,
+      LONG_RUNNING_PERMISSION,
+    );
+
+    // Long running snaps have timeouts disabled
+    if (isLongRunning) {
+      return promise;
+    }
```
Comment (rekmarks) [inline packages/controllers/src/snaps/SnapController.ts:1778]: I'm thinking should probably just leave this alone for now with the intention to eventually support some kind of "long-running worker" API and add some kind of timeout here at that point. We need to make sure to warn about this development in big red letters in the docs, but other than that, this is OK.

File: packages/controllers/src/snaps/SnapController.test.ts [hunk 2]
```diff
@@ -342,7 +355,19 @@ describe('SnapController', () => {
...
+    const messenger = getSnapControllerMessenger();
+    jest.spyOn(messenger, 'call').mockImplementation((method, ...args) => {
+      if (
+        method === 'PermissionController:hasPermission' &&
+        args[1] === LONG_RUNNING_PERMISSION
+      ) {
+        return false;
+      }
+      return true;
+    });
```
Comment (rekmarks) [inline packages/controllers/src/snaps/SnapController.test.ts:367]: This boilerplate is really blowing up this test file. Is there a neat way to encapsulate this within `getSnapControllerMessenger()`?

**Review by rekmarks — COMMENTED** [review]
Adding some requested changes here.

File: packages/controllers/src/snaps/SnapController.ts (hunk 1 above)
Comment (FrederikBolding) [reply packages/controllers/src/snaps/SnapController.ts:1778]: 👍

File: packages/controllers/src/snaps/SnapController.test.ts (hunk 2 above)
Comment (FrederikBolding) [reply packages/controllers/src/snaps/SnapController.test.ts]: I don't know how I feel about it, but I have added a default spy to `getSnapControllerMessenger`, this simplifies the tests a bit.

Comment (rekmarks) [reply packages/controllers/src/snaps/SnapController.test.ts]: I like what you did :+1:

**Review by hmalik88 — APPROVED** [review]
LGTM.

**Pull request merged**